pkgs/racket-doc/scribblings/reference/equality.scrbl

```diff
@@ -11,16 +11,46 @@ preferred for most use cases.
 
 @defproc[(equal? [v1 any/c] [v2 any/c]) boolean?]{
 
- Two values are @racket[equal?] if and only if they are @racket[eqv?],
- unless otherwise specified for a particular datatype.
-
- Datatypes with further specification of @racket[equal?] include
- strings, byte strings, pairs, mutable pairs, vectors, boxes, hash
- tables, and inspectable structures. In the last six cases, equality
- is recursively defined; if both @racket[v1] and @racket[v2] contain
- reference cycles, they are equal when the infinite unfoldings of the
- values would be equal. See also @racket[gen:equal+hash] and
- @racket[prop:impersonator-of].
+ Determines whether @racket[v1] is ``equal to'' @racket[v2], in a manner that
+ depends on the datatype of @racket[v1] and @racket[v2].
+
+ Different datatypes have different implementations of @racket[equal?], but the
+ golden rule uniting them is that @deftech{equal means interchangeable} --- if
```

Comment:
This use of the word "interchangable" is fraught for people with a PL background. Equal? values aren't interchangable in this sense when you get technical.

Reply:
How can this be a “golden rule” if `gen:equal+hash` doesn’t enforce any of these attributes?

Reply:
The problem is that this is a higher-order property, so it's not really enforce-able at the definition site (you would need a verifier for that, and yet it still won't work for all situations. These problems are in general undecidable). We can imagine dynamically checking some properties like symmetry by, for every call of `(equal? a b)`, calling `(equal? b a)` to see if they result in the same thing, but that's also prohibitively expensive.

Reply:
But I think it's accurate to say that "equality must be equivalence relation" is a golden rule. Everything breaks otherwise.

Reply:
I think of the golden rule as one that can be broken, but it isn't right Right Thing to break it. :) 

That said, it does usually refer to a specific rule so maybe another phrase is better?

Reply:
“Equal means interchangeable” is a curious phrase, because a) it doesn’t illuminate the meaning of “equal” — it just invokes a different word (“interchangeable”) that is no more concrete, and b) if the essential notion of “equal” is “equivalence relation”, fair enough, but there are plenty of equivalence relations that do not produce intuitively “interchangeable” values (for instance `abs`)

Reply:
I think another thing that @rfindler was trying to point out regarding "Equal means interchangeable" is that it's technically false. `(+ 1 1)` is `equal?` to `2`, but they are not interchangeable in some contexts. E.g., `(quote (+ 1 1))` and `(quote 2)`.

Reply:
I was thinking of something slightly different. I think that the quote thing isn't a big issue as you can just rule out contexts inside quotes. But if you have somewhere `(define x (list 1 2))` even if that exact `x` is in scope, you cannot replace it with `(list 1 2)` because these can be distinguished (using `eq?`). But more importantly, I don't think that this kind of thing is really what is important about `equal?` anyway (I put what I think is important about it in a different comment around here somewhere...)